### PR TITLE
Keep web contents alive and wait for media files to be added

### PR DESCRIPTION
**Pull request by sangwoo108**
There could be additional media files added to a page after we already executed detection script. In order to detect them, attach MutationObserver via PlaylistRenderFrameObserver

Resolves https://github.com/brave/brave-browser/issues/29768
Resolves https://github.com/brave/brave-browser/issues/29488

## Submitter Checklist:

- [ ] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### Automated
PlaylistDownloadRequestManagerBrowserTest.DynamicallyAddedMedia

### note
locally tested and verified that it's working - @deeppandya is going to follow up with this.

### CODE REVIEW

File: components/playlist/renderer/playlist_render_frame_observer.cc
Comment (petemill) [inline components/playlist/renderer/playlist_render_frame_observer.cc:102]: I'm confused about having both `setInterval` here and `setTimeout`:
1. `setInterval` with no `delay` parameter will cause it to call `findNewMEdiaAndObserverMutation` in a loop with no delay between execution (`delay` defaults to `0`).
2. Then `setTimeout` is called with a delay of 20 seconds. In 20 seconds, the loop started in 1. will be cancelled, and the `findNewMediaAndObserveMutation` function will be scheduled to run 1 more time when the thread is next "idle".

Is this because we want to try to catch the src changing and we want the first version? Why not have a delay on the `setInterval` and just check periodically, e.g. every 1 second? I'm worried about performance here. 

Another alternative you could do is have a MutationObserver on the main document, and get notified of new video elements, and/or monkey-patch `document.createElement` and handle when the element is `video` or `audio`? I think this may be what iOS does?

File: components/playlist/browser/playlist_service.cc
Comment (fallaciousreasoning) [inline components/playlist/browser/playlist_service.cc:328]: Should we still fire the listeners if there are no items (i.e. the site removed all the videos).

File: components/playlist/renderer/BUILD.gn
Comment (fallaciousreasoning) [inline components/playlist/renderer/BUILD.gn:30]: Nit: Duplicate dep

File: components/playlist/renderer/playlist_render_frame_observer.cc
Comment (fallaciousreasoning) [inline components/playlist/renderer/playlist_render_frame_observer.cc:81]:
```suggestion
      const mutationSources = new Set();
```

File: components/playlist/renderer/playlist_render_frame_observer.cc
Comment (fallaciousreasoning) [inline components/playlist/renderer/playlist_render_frame_observer.cc:89]: FWIW, the same mutation observer can observe multiple elements - you can call `observe` as many times as you need on different elements. (https://stackoverflow.com/questions/30807846/can-a-single-mutationobserver-object-observe-multiple-targets)

File: components/playlist/renderer/playlist_render_frame_observer.cc
Comment (fallaciousreasoning) [inline components/playlist/renderer/playlist_render_frame_observer.cc:90]: If we're not sharing the observer between multiple elements we know exactly what changed here so we could (optionally) do this instead:

```suggestion
                pl_worker.onMediaUpdated(mediaNode.src);
```

File: components/playlist/renderer/playlist_render_frame_observer.cc
Comment (fallaciousreasoning) [reply components/playlist/renderer/playlist_render_frame_observer.cc:102]: I read this as "Let's poll for 20 seconds and see if we get a video". If we didn't get anything enqueue an idle callback, which will hopefully fire when the document finishes changing things around.

:+1: to the 1 second interval

I don't think patching `document.createElement` will work for us, because we'll miss elements created via `innerHTML` and `document.write` (and whatever other crazy ways you can create elements) but we could definitely use a mutation observer on the body:

This should behave identically, but only do the find in ticks where the DOM changed - dunno if it's better than just checking every second though :laughing: 
```js
const mutationObserver = new MutationObserver(e => {
    // todo: be cleverer here and only do this if video/audio elements were added/removed/src changed
    findNewMediaAndObserveMutation()
})
mutationObserver.observer(document.body, { attributeFilter: ['src'], childList: true, subtree: true })

setTimeout(() => {
    mutationObserver.disconnect();
    window.requestIdleCallback(findNewMediaAndObserveMutation);
}, 20000);
```

Comment (fallaciousreasoning) [reply components/playlist/renderer/playlist_render_frame_observer.cc:102]: Also, out of curiosity, what does the `pollingIntervalId` do as an argument to `setTimeout`? I've never seen it before

```js
setTimeout(() => ..., 20000, pollingIntervalId)
```

File: browser/playlist/test/playlist_download_request_manager_browsertests.cc
Comment (fallaciousreasoning) [inline browser/playlist/test/playlist_download_request_manager_browsertests.cc:105]: nit:
```suggestion
    // Start server: ASSERT_TRUE() has its own return type internally, thus
```

Comment (fallaciousreasoning) [reply browser/playlist/test/playlist_download_request_manager_browsertests.cc:105]: Cool! I didn't think of doing this! That's a cool way of being able to `ASSERT` anywhere!

**Review by fallaciousreasoning — APPROVED** [review]
Looks pretty good to me @sangwoo108 - few questions about the approach to detecting new videos/audio but nothing too major.

File: components/playlist/renderer/playlist_render_frame_observer.cc
Comment (sangwoo108) [reply components/playlist/renderer/playlist_render_frame_observer.cc:102]: > setInterval with no delay parameter will cause it to call findNewMEdiaAndObserverMutation in a loop with no delay between execution (delay defaults to 0).

> setTimeout(() => ..., 20000, pollingIntervalId)

Ah, my bad. I made a mistake during refactoring.

---

> setInterval and just check periodically, e.g. every 1 second? I'm worried about performance here.

>  I understand this: We poll for 20 seconds and see if we get a video. If we didn't get anything enqueue an idle callback, 

Yeah, the overall strategy is from cosmetic_js_handler.cc, which is switching between polling and idle-time detection. So What I try to do is polling for the first seconds, then turn it to idle-time detection. The reason why polling first is that it's more likely for videos to be attached in the beginning. 

Does polling is better than rIF in terms of performance?

---

And one more thing, @diracdeltas and I was talking about the performance problem of `MutationObserver` and I guess we can look into it later as this scrpit will be run only on the background/copied web conetents for now. - https://github.com/brave/brave-browser/issues/29488#issuecomment-1502666818

And there's a native API which acts just like `MutationObserver`, so I'm going to try it, before too late.

File: components/playlist/renderer/DEPS
Comment (goodov) [inline components/playlist/renderer/DEPS:11]: duplicated deps

File: components/playlist/renderer/playlist_js_handler.cc
Comment (goodov) [inline components/playlist/renderer/playlist_js_handler.cc:95]: I haven't found the right answer if it's safe to use `Unretained` here, but I see Chromium uses `WeakPtr` in many places.

File: components/playlist/renderer/playlist_render_frame_observer.cc
Comment (sangwoo108) [reply components/playlist/renderer/playlist_render_frame_observer.cc:102]: @fallaciousreasoning , I'd like to ask something about your suggestion

```
const mutationObserver = new MutationObserver(e => {
    // todo: be cleverer here and only do this if video/audio elements were added/removed/src changed
    findNewMediaAndObserveMutation()
})
mutationObserver.observer(document.body, { attributeFilter: ['src'], childList: true, subtree: true })
```

Does this mean that one `MutationObserver` will observe the entire DOM tree from the document.body as a root? I'm concerned if it'll cause performance issue mentioned [here](https://github.com/brave/brave-browser/issues/29488#issuecomment-1502666818). And that was the initial idea why I made a multiple mutation observer that observes only "src" attribute. And IIRC, `document.body` wasn't valid at the time this code was executed 😥

File: components/playlist/browser/playlist_service.cc
Comment (sangwoo108) [reply components/playlist/browser/playlist_service.cc:328]: For now, I guess we don't need to, as this event is only about dynamically added files. On the other hands, if the listeners want to make it sure, they can explicitly call another method, then they'll get an empty list.

File: components/playlist/renderer/playlist_js_handler.cc
Comment (sangwoo108) [reply components/playlist/renderer/playlist_js_handler.cc]: Oops, my bad. Changed to use Weakptr.

File: components/playlist/renderer/playlist_render_frame_observer.cc
Comment (sangwoo108) [reply components/playlist/renderer/playlist_render_frame_observer.cc]: I was confused between js and c++ as for constness again haha😄

File: browser/playlist/test/playlist_download_request_manager_browsertests.cc
Comment (goodov) [inline browser/playlist/test/playlist_download_request_manager_browsertests.cc:486]: 1. can we do `EXPECT_CALL().WillOnce(... { run_loop_Quit(); })` instead of `ON_CALL + EXPECT_CALL`?
2. pls replace `base::BindLambdaForTesting` with `MockCallback` (`base/test/mock_callback.h`). It's a bad practice to ASSERT/EXPECT in a callback that might not be called at all.

File: browser/playlist/test/playlist_download_request_manager_browsertests.cc
Comment (goodov) [inline browser/playlist/test/playlist_download_request_manager_browsertests.cc:115]: `Serve` can be `static`. `base::Unretained` can be removed.

File: components/playlist/browser/playlist_download_request_manager.h
Comment (goodov) [inline components/playlist/browser/playlist_download_request_manager.h:76]: should this really be non-const?

**Review by goodov — APPROVED** [review]
lgtm with few nits

File: android/java/org/chromium/chrome/browser/playlist/PlaylistServiceObserverImpl.java
Comment (deeppandya) [inline android/java/org/chromium/chrome/browser/playlist/PlaylistServiceObserverImpl.java:42]: We need to handle this callback in a different branch for android specific changes.

**Review by deeppandya — APPROVED** [review]
Android changes LGTM. verified the changes with the build.

File: browser/playlist/test/playlist_download_request_manager_browsertests.cc
Comment (sangwoo108) [reply browser/playlist/test/playlist_download_request_manager_browsertests.cc]: Thanks a lot! Didn't know those.

File: components/playlist/browser/playlist_download_request_manager.h
Comment (sangwoo108) [reply components/playlist/browser/playlist_download_request_manager.h]: No, changed to const method.

File: components/playlist/renderer/playlist_render_frame_observer.cc
Comment (fallaciousreasoning) [reply components/playlist/renderer/playlist_render_frame_observer.cc]: Yeah, if we're concerned about perf the polling approach might be better :smile: Optionally, we could still share the mutation observer if we wanted (and just call `.observe` on all the video/audio elements).

**Comment by sangwoo108** [comment]
@fallaciousreasoning @petemill Do you mind taking another look?

**Review by simonhong — APPROVED** [review]
components/playlist/browser ++

**Comment by sangwoo108** [comment]
Hi, @petemill @fallaciousreasoning . I was waiting for the second round, just in case. But it looks like we don't have much disagreement on this. Please let me merge this as this PR got the minimum approval from reviewers, and let me know if there's anything to discuss anytime.

File: components/playlist/browser/playlist_render_frame_browser_client.cc
Comment (bridiver) [inline components/playlist/browser/playlist_render_frame_browser_client.cc:12]: This naming convention is not correct. `*BrowserClient` is used either for direct subclasses of ContentBrowserClient or for "delegates" that implement some of those methods and are called from the main Brave/ChromeContentBrowserClient. Something like `PlaylistMediaHandler` would be more appropriate here

File: components/playlist/browser/playlist_render_frame_browser_client.cc
Comment (bridiver) [inline components/playlist/browser/playlist_render_frame_browser_client.cc:14]: `WeakPtr` is not necessary here, keyed services will always outlive all render frames

File: components/playlist/browser/playlist_service.cc
Comment (bridiver) [inline components/playlist/browser/playlist_service.cc:380]: this is not the correct way to expose this publicly. The class should subclass `public base::SupportsWeakPtr<PlaylistService>` which exposes `AsWeakPtr`

Comment (sangwoo108) [reply components/playlist/browser/playlist_service.cc:380]: I think Upstream is going to remove SupportsWeakPtr. https://bugs.chromium.org/p/chromium/issues/detail?id=647430&q=SupportsWeakPtr&can=2

File: components/playlist/renderer/playlist_js_handler.cc
Comment (bridiver) [inline components/playlist/renderer/playlist_js_handler.cc:76]: It is not ok to add things to the global context in the main world

Comment (bridiver) [reply components/playlist/renderer/playlist_js_handler.cc:76]: it's also not ok to add JS apis to the main world. Did this get a security review?

File: components/playlist/renderer/playlist_js_handler.cc
Comment (bridiver) [inline components/playlist/renderer/playlist_js_handler.cc:93]: what prevents a site from calling this to automatically download videos to playlist? On top of the issues mentioned here https://github.com/brave/brave-core/pull/17994/files#r1369288074 this also doesn't even check for user activation or any kind of throttling

Comment (bridiver) [reply components/playlist/renderer/playlist_js_handler.cc:93]: also what prevents a site from including a bunch of hidden media elements that might contain objectionable/illegal content?

File: components/playlist/renderer/playlist_js_handler.cc
Comment (bridiver) [reply components/playlist/renderer/playlist_js_handler.cc:76]: I guess this is an isolated world. There was a mention of a button "on page" so I was looking to see what that button did, but apparently the download is triggered from browser UI

Comment (sangwoo108) [reply components/playlist/renderer/playlist_js_handler.cc:76]: > It is not ok to add things to the global context in the main world
> it's also not ok to add JS apis to the main world. Did this get a security review?

No. I thought we were supposed to have a final review before the Android release(enbaling the feature), but we're having one for desktop first.

We needed MutationObserver for a site which updates video src later on(I think it was TED.) So I mimicked what cosmetic filter js handler does. And this mutation observer will be attached only to the offscreen contents which mirrors an actual tab's url.

---

If this shouldn't be done in this way, what should it be like?

File: components/playlist/renderer/playlist_js_handler.cc
Comment (sangwoo108) [reply components/playlist/renderer/playlist_js_handler.cc:93]: We don't have that mechanism. Instead we're showing up chooser so that user can confirm the items to be added.

**Pull request merged**
